### PR TITLE
NEAR support

### PR DESCRIPTION
**Pull request by evaporei**
This PR includes:

- New manifest validation for NEAR subgraphs
- `codegen` support for NEAR
- `build` support for NEAR
- `deploy` support for NEAR

I've created a `Protocol` class and respective classes for necessary implementations to each protocol (codegen, ABI, etc).

For the circular dependencies commits, I've used this tool to debug it: https://www.npmjs.com/package/madge.

It still needs more manual testing (running the commands against ethereum and near subgraphs) just to make sure everything is working 🙂  (me and @maoueh are looking at this).

What's missing for the future?

- `graph init`
- Dynamic data sources / `templates` (maybe one day, if necessary)
- `functionCallHandlers`

Related Issue: https://github.com/graphprotocol/graph-cli/issues/734

### CODE REVIEW

File: src/protocols/near/subgraph.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  validateContractAddresses() {
+    // Reference: https://docs.near.org/docs/concepts/account#account-id-rules
+    const MINIMUM_ACCOUNT_ID_LENGTH = 2
+    const MAXIMUM_ACCOUNT_ID_LENGTH = 64
+    const validateLength = accountId =>
+      accountId.length >= MINIMUM_ACCOUNT_ID_LENGTH &&
+      accountId.length <= MAXIMUM_ACCOUNT_ID_LENGTH
+    const nearAccountIdPattern = /^(([a-z\d]+[\-_])*[a-z\d]+\.)*([a-z\d]+[\-_])*[a-z\d]+$/
```
[the remark this answers is not in the document]
Comment (otaviopace) [reply src/protocols/near/subgraph.js:22]: That's from their docs haha, as linked above.

File: manifest-schema.graphql [hunk 2]
```diff
@@ -17,14 +22,27 @@ type EthereumContractDataSource {
...
+}
+
 type EthereumContractSource {
   address: String
   abi: String!
   startBlock: BigInt
 }
 
+type NearContractSource {
+  address: String
```
Comment (leoyvens) [inline manifest-schema.graphql:40]:
```suggestion
  account: String
```

File: manifest-schema.graphql [hunk 3]
```diff
@@ -35,6 +53,16 @@ type EthereumContractMapping {
...
 }
 
+type NearContractMapping {
+  apiVersion: String!
+  language: String!
+  file: File!
+  entities: [String!]!
+  blockHandlers: [NearBlockHandler!]
+  receiptHandlers: [NearReceiptHandler!]
+  functionCallHandlers: [NearFunctionCallHandler!]
```
Comment (leoyvens) [inline manifest-schema.graphql:63]: Lets comment out `functionCallHandlers` for now.

**Comment by otaviopace** [comment]
@leoyvens I've made all of the suggested changes in the last three commits, feel free to take a look 😊

File: manifest-schema.graphql (hunk 3 above)
Comment (otaviopace) [reply manifest-schema.graphql]: ⛔   **r e m o v e d** ⛔

File: src/protocols/near/codegen/template.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    return tsCodegen.staticMethod(
+      'create',
+      [tsCodegen.param('accountId', tsCodegen.namedType('near.AccountId'))],
+      tsCodegen.namedType('void'),
+      `
+      DataSourceTemplate.create('${name}', [accountId])
+      `,
+    )
+  }
```
[the remark this answers is not in the document]
Comment (otaviopace) [reply src/protocols/near/codegen/template.js]: They're removed (from the PR description as well).

**Review by leoyvens — APPROVED** [review]
Lets not squash on merging so we can keep the template support in the history.

**Pull request merged**